### ISSUE TITLE
twitter 401 Authorization Required error

### ISSUE THREAD

**Issue by jdragoo** [cantino/huginn#988]
I have seen a few tickets on this issue, but cannot seem to solve it using the solutions that were found at the time.  I have created my Twitter app and setup my keys in the .env file.  I have also made sure the callback url is setup according to the documentation, but I still get the 401 Auth error on the services page when I try to link it.  My server is running in AWS on ubuntu if that makes any difference.

**Comment by dsander** [comment]
Hi @jdragoo,

can you please post the complete exception include the backtrace and the obscured twitter oauth configution of your `.env` file?

**Comment by jdragoo** [comment]
This is the backtrace I see on the webpage after the failure.  Below is the section of the .env file with the Twitter OAUTH info removed.


oauth (0.4.7) lib/oauth/consumer.rb:216:in `token_request'
oauth (0.4.7) lib/oauth/consumer.rb:136:in `get_request_token'
omniauth-oauth (1.0.1) lib/omniauth/strategies/oauth.rb:29:in `request_phase'
omniauth-twitter (1.0.1) lib/omniauth/strategies/twitter.rb:60:in `request_phase'
omniauth (1.2.2) lib/omniauth/strategy.rb:215:in `request_call'
omniauth (1.2.2) lib/omniauth/strategy.rb:183:in `call!'
omniauth (1.2.2) lib/omniauth/strategy.rb:164:in `call'
warden (1.2.3) lib/warden/manager.rb:35:in `block in call'
warden (1.2.3) lib/warden/manager.rb:34:in `catch'
warden (1.2.3) lib/warden/manager.rb:34:in `call'
rack (1.6.4) lib/rack/etag.rb:24:in `call'
rack (1.6.4) lib/rack/conditionalget.rb:25:in `call'
rack (1.6.4) lib/rack/head.rb:13:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/flash.rb:260:in `call'
rack (1.6.4) lib/rack/session/abstract/id.rb:225:in `context'
rack (1.6.4) lib/rack/session/abstract/id.rb:220:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/cookies.rb:560:in `call'
activerecord (4.2.2) lib/active_record/query_cache.rb:36:in `call'
activerecord (4.2.2) lib/active_record/connection_adapters/abstract/connection_pool.rb:649:in `call'
activerecord (4.2.2) lib/active_record/migration.rb:378:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
activesupport (4.2.2) lib/active_support/callbacks.rb:88:in `call'
activesupport (4.2.2) lib/active_support/callbacks.rb:88:in `_run_callbacks'
activesupport (4.2.2) lib/active_support/callbacks.rb:776:in `_run_call_callbacks'
activesupport (4.2.2) lib/active_support/callbacks.rb:81:in `run_callbacks'
actionpack (4.2.2) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/reloader.rb:73:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
better_errors (1.1.0) lib/better_errors/middleware.rb:58:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
railties (4.2.2) lib/rails/rack/logger.rb:38:in `call_app'
railties (4.2.2) lib/rails/rack/logger.rb:20:in `block in call'
activesupport (4.2.2) lib/active_support/tagged_logging.rb:68:in `block in tagged'
activesupport (4.2.2) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (4.2.2) lib/active_support/tagged_logging.rb:68:in `tagged'
railties (4.2.2) lib/rails/rack/logger.rb:20:in `call'
config/initializers/silence_worker_status_logger.rb:5:in `call_with_silence_worker_status'
quiet_assets (1.1.0) lib/quiet_assets.rb:27:in `call_with_quiet_assets'
actionpack (4.2.2) lib/action_dispatch/middleware/request_id.rb:21:in `call'
rack (1.6.4) lib/rack/methodoverride.rb:22:in `call'
rack (1.6.4) lib/rack/runtime.rb:18:in `call'
activesupport (4.2.2) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
rack (1.6.4) lib/rack/lock.rb:17:in `call'
actionpack (4.2.2) lib/action_dispatch/middleware/static.rb:113:in `call'
rack (1.6.4) lib/rack/sendfile.rb:113:in `call'
railties (4.2.2) lib/rails/engine.rb:518:in `call'
railties (4.2.2) lib/rails/application.rb:164:in `call'
rack (1.6.4) lib/rack/lock.rb:17:in `call'
rack (1.6.4) lib/rack/content_length.rb:15:in `call'
rack (1.6.4) lib/rack/handler/webrick.rb:88:in `service'
/usr/local/lib/ruby/2.2.0/webrick/httpserver.rb:138:in `service'
/usr/local/lib/ruby/2.2.0/webrick/httpserver.rb:94:in `run'
/usr/local/lib/ruby/2.2.0/webrick/server.rb:294:in `block in start_thread'


########################################################################################################
#    OAuth Configuration                                                                               #
#  More information at the wiki: https://github.com/cantino/huginn/wiki/Configuring-OAuth-applications #
########################################################################################################

TWITTER_OAUTH_KEY=---------------------
TWITTER_OAUTH_SECRET=----------------
TWITTER_CONSUMER_KEY=
TWITTER_CONSUMER_SECRET=

**Comment by cantino** [comment]
I assume you've made a Service in the UI as well?  (I forgot we still had the consumer key and secret here, you shouldn't need those, and I just removed them in #989.)

**Comment by jdragoo** [comment]
I get the error when making the service in the UI.  When I click the auth button under services it takes me to the error attached in the screen shot.  I tried it with the consumer keys as well just to make sure and of course still the same error.

![screen shot 2015-08-24 at 10 16 47 pm](https://cloud.githubusercontent.com/assets/13945702/9457182/f89f60dc-4aad-11e5-9ab4-c9fc13083d7a.png)

**Comment by cantino** [comment]
Oh, I see.  What did you put as the callback URL when you registered with Twitter?  When I've had trouble in the past, this is often the problem.

**Comment by jdragoo** [comment]
http://huginn.jasonremote.com:3000/auth/twitter/callback

**Comment by dsander** [comment]
That looks correct, can you try to log out of your twitter account before you click on the "authenticate with twitter" button?

**Comment by jdragoo** [comment]
I tried that with no luck.  Do we think this is a problem with the Twitter app I created?  I have created a few and given them full access to everything and still have trouble.

**Comment by dsander** [comment]
I don't know, maybe try to delete your cookies? I remember having this issues once, and think I fixed it by signing out and back into huginn/twitter.

**Comment by cantino** [comment]
I agree about trying log in / out in both systems.

Also make sure you don't have any spaces or anything in the TWITTER_OAUTH_KEY or TWITTER_OAUTH_SECRET.

**Comment by Ishfal** [comment]
I've had this same problem for a while -- just tried wiping cookies, logging out, etc., without impact. I'm running on Heroku, and observed that the Huginn logs read as follows:

2015-08-28T01:51:13.941913+00:00 app[web.1]: [71bc5dc0-1609-47f3-9997-0fda9871bb3f] Started GET "/auth/twitter" for xx.xx.xx.xx at 2015-08-28 01:51:13 +0000
2015-08-28T01:51:14.045186+00:00 app[web.1]: [71bc5dc0-1609-47f3-9997-0fda9871bb3f]
2015-08-28T01:51:14.045193+00:00 app[web.1]: OAuth::Unauthorized (401 Authorization Required):
2015-08-28T01:51:14.045196+00:00 app[web.1]:   config/initializers/silence_worker_status_logger.rb:5:in `call_with_silence_worker_status'
2015-08-28T01:51:14.045198+00:00 app[web.1]:
2015-08-28T01:51:14.045199+00:00 app[web.1]:
2015-08-28T01:51:13.943801+00:00 app[web.1]: I, [2015-08-28T01:51:13.943650 #20]  INFO -- omniauth: (twitter) Request phase initiated.

At one point, I theorized that perhaps for some reason the Twitter Oauth tokens aren't being provided, even though they are configured?

**Comment by cantino** [comment]
I assume that other changes to your .env are being picked up?  Since you're on Heroku, are you using the .env, or are you using `heroku config:set VARIABLE=value`?

**Comment by Ishfal** [comment]
Yes, I had been setting via the Config variables. I had also tried putting them in as Credentials as well, but to no effect. 

On August 28, 2015 8:13:59 AM PDT, Andrew Cantino <notifications@github.com> wrote:
>I assume that other changes to your .env are being picked up?  Since
>you're on Heroku, are you using the .env, or are you using `heroku
>config:set VARIABLE=value`?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/cantino/huginn/issues/988#issuecomment-135800815

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by cantino** [comment]
Do you have a `.env` file too?  They might be conflicting.

**Comment by Ishfal** [comment]
Nope, no .env file

**Comment by cantino** [comment]
I'm kind of at a loss.  Random thoughts:

* Could there be extra quotes (`"`) around the code in Heroku's settings?
* Do other settings work?
* Do other services work?

**Comment by Ishfal** [comment]
Good call, let me try other services and report back. That might help to isolate the issue. 

Am I interpreting the logs correctly that it is failing to provide/read the right OAuth info?

On August 28, 2015 9:38:09 PM PDT, Andrew Cantino <notifications@github.com> wrote:
>I'm kind of at a loss.  Random thoughts:
>
>* Could there be extra quotes (`"`) around the code in Heroku's
>settings?
>* Do other settings work?
>* Do other services work?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/cantino/huginn/issues/988#issuecomment-135940751

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by cantino** [comment]
I'm not entirely sure.  Going to `/auth/twitter` should redirect you to Twitter's site for confirmation, then back to `/auth/twitter/callback`, which is the callback URL you should have given them.

**Comment by jdragoo** [comment]
So I have tried other services without a problem since that seems to be the next step.  The only one not working is Twitter.  I am in the process of setting it up from scratch on a test server to see if I get the same problem or if it works and could be related to the environment.

**Comment by cantino** [comment]
What callback URL did you give Twitter when you set it up?

**Comment by Ishfal** [comment]
I made sure to populate the App's callback on Twitter with the http://[NAME].herokuapp.com/auth/twitter/callback URL. I actually created two separate apps to test it (in case the first was flaky somehow), and neither seemed to make a difference.

**Comment by cantino** [comment]
http_s_://, right?

**Comment by Ishfal** [comment]
Apologies, yes, that is correct.

**Comment by cantino** [comment]
Here's what my app looks like.  Anything different?

<img width="743" alt="screen shot 2015-08-31 at 8 49 37 pm" src="https://cloud.githubusercontent.com/assets/83835/9595896/d836ce06-5021-11e5-96ba-ae74a2a92078.png">

**Comment by cantino** [comment]
Actually, never mind, that one is wrong.

**Comment by cantino** [comment]
That all seems right...  `/auth/twitter/callback` is what I use.  I wish we knew more about the error.  You could try running `heroku run rails console` and looking at `ENV` and making sure it looks right.

**Comment by Ishfal** [comment]
Vs. the App config above, mine was set as 'Read-Write' access and 'Sign-in' as 'Yes'. I changed both but that didn't seem to have any impact.

Just pulled the ENV and can confirm that shows all of the correct Twitter Oauth info.

**Comment by cantino** [comment]
That's very strange.  @dsander, @knu, anyone have any other ideas?

**Comment by knu** [comment]
You probably need to track redirects using the browser's developer console.

**Comment by jdragoo** [comment]
I tried tracking redirects and I see a 500 Internal Server error when it calls http://site.com:3000/auth/twitter

**Comment by knu** [comment]
You mean literally?  With no `access_token` parameter?

**Comment by jdragoo** [comment]
I'm not sure what you mean about the access_token parameter.  I just used firebug to follow where it was going when I clicked on the Auth with twitter button and it showed the final redirect back to my site in red with a 500 Internal Server error.

**Comment by knu** [comment]
Oh wait, that's not the callback.  You aren't redirected to Twitter.com.  I have no idea at the moment, then.

**Comment by cantino** [comment]
In Chrome, in the Network tab of the Developer Tools, you checked the `[ ] Preserve log` checkbox, right?

**Comment by jdragoo** [comment]
It is checked.  Below is an attached screen shot with the red 500 error at the bottom that I get when clicking the Auth with Twitter button

![screen shot 2015-09-01 at 11 08 45 pm](https://cloud.githubusercontent.com/assets/13945702/9622260/a103ab42-50fe-11e5-91f2-f5996a80e0f8.png)

**Comment by cantino** [comment]
Very odd.  Maybe it's not sending a param it needs?

**Comment by Ishfal** [comment]
I get more or less the same thing, which unfortunately doesn't really give any clues as to what the issue is.

**Comment by cantino** [comment]
It does seem like we have a Heroku bug.  Any chance you could give me access to your Heroku instance to poke around?  Email me at andrew [at] andrewcantino.com if you'd be open to that.

**Comment by jdragoo** [comment]
So I have installed Huginn from scratch on three different VM's both Ubuntu and Centos on and off AWS to see if there is anything I could find.  The final install always gives me this error when setting up the Twitter app.  The other services I configure seem to be running fine though.  Is it possible there is some prerequisite that we may be missing for Twitter to work?

**Comment by cantino** [comment]
Got your email-- I'll dig into it today or tomorrow.

**Comment by cantino** [comment]
See the email I just sent.

For everyone else: I figured out the issue-- @Ishfal had understandably gotten confused by the `TWITTER_CONSUMER_KEY` and `TWITTER_CONSUMER_SECRET` ENV variables that were in the .env.example file until I removed them in #989.  He had gotten the oauth application secret and token confused with these and had them switched.  Hopefully with their removal, this won't confuse anyone else.

**Comment by Ishfal** [comment]
Got it, thanks @cantino! So, for other users' benefit, the only info needed is what Twitter calls the 'Consumer Key (API Key' and the 'Consumer Secret (API Secret)', in the Apps interface (plus the right callback URL, etc.). Twitter's 'Access Token' and 'Access Token Secret' are not needed. Am I stating that correctly?

@jdragoo, does this solve your problem if you switch those?

**Comment by jdragoo** [comment]
Thanks @cantino, that was my problem as well.  I had the keys crossed.  Once I switched them it let me connect no problem.

**Comment by cantino** [comment]
Awesome, I'm glad it's working for both of you!  I'm going to close this issue.  Please open a new one if you run into anything else. :smile:

**Comment by gowsri11** [comment]
I am getting the same error and none of the above solutions work :(

**Comment by bmorris-ocsd** [comment]
> I am getting the same error and none of the above solutions work :(

Same. This error is still occuring. I've been fighting this problem for a couple of days now. I started with the quickstart, that yielded the problem initially. I spent an entire day trying different things. This morning I came in and decided to follow https://github.com/huginn/huginn/blob/master/doc/manual/installation.md
and STILL no joy. Ruby is -not- my language and the Huginn framework has a crazy amount of gems being used and integrated making it a really complex beast to debug for someone whose language isn't Ruby, much worse, Ruby on Rails. If I run in development, I see the error as a 401 (At one point I saw a 403, but it might have just been a callback URL I adjusted on the app page in Twitter to cause that), which is Unauthrorized. However, if I bring up the developer console in Chrome, I see that it's returning to http://huginn/services with a 500 Internal Server error. Even in production mode, and hitting http://huginn/auth/twitter seems to yield a 500 Internal Server error.

This either seems like a major regression or the bug was not really actually fixed.
